tests/test_cli.py

```diff
@@ -0,0 +1,297 @@
+
+import argparse
+import io
+import os
+import tempfile
+from unittest.mock import AsyncMock, MagicMock, mock_open, patch
+
+import pytest
+
+from pybricksdev.cli import Push, Tool
+
+
+class TestTool:
+    """Tests for the base Tool class."""
+
+    def test_is_abstract(self):
+        """Test that Tool is an abstract base class."""
+        with pytest.raises(TypeError):
+            Tool()
+
+
+class TestPush:
+    """Tests for the Push command."""
+
+    def test_add_parser(self):
+        """Test that the parser is set up correctly."""
+        # Create a subparsers object
+        parser = argparse.ArgumentParser()
+        subparsers = parser.add_subparsers()
+
+        # Add the push command
+        push = Push()
+        push.add_parser(subparsers)
+
+        # Verify the parser was created with correct arguments
+        assert "push" in subparsers.choices
+        parser = subparsers.choices["push"]
+        assert parser.tool is push
+
+        # Test that required arguments are present
+        mock_file = mock_open(read_data="print('test')")
+        mock_file.return_value.name = "test.py"
+        with patch("builtins.open", mock_file):
+            args = parser.parse_args(["ble", "test.py"])
+            assert args.conntype == "ble"
+            assert args.file.name == "test.py"
+            assert args.name is None
+
+        # Test with optional name argument
+        mock_file = mock_open(read_data="print('test')")
+        mock_file.return_value.name = "test.py"
+        with patch("builtins.open", mock_file):
+            args = parser.parse_args(["ble", "test.py", "-n", "MyHub"])
+            assert args.name == "MyHub"
+
+        # Test that invalid connection type is rejected
+        with pytest.raises(SystemExit):
+            parser.parse_args(["invalid", "test.py"])
+
+    @pytest.mark.asyncio
+    async def test_push_ble(self):
+        """Test running the push command with BLE connection."""
+        # Create a mock hub
+        mock_hub = AsyncMock()
+        mock_hub._mpy_abi_version = 6
+        mock_hub.download_user_program = AsyncMock()
+
+        # Create a temporary file
+        with tempfile.NamedTemporaryFile(suffix=".py", mode="w+", delete=False) as temp:
+            temp.write("print('test')")
+            temp_path = temp.name
+
+        try:
+            # Create args
+            args = argparse.Namespace(
+                conntype="ble",
+                file=open(temp_path, "r"),
+                name="MyHub",
+            )
+
+            # Mock the compilation
+            mock_mpy = MagicMock()
+            mock_mpy.__bytes__ = lambda self: b"compiled code"
+
+            # Mock the hub creation
+            with patch(
+                "pybricksdev.connections.pybricks.PybricksHubBLE", return_value=mock_hub
+            ) as mock_hub_class:
+                with patch("pybricksdev.ble.find_device", return_value="mock_device"):
+                    with patch(
```

Comment:
This is getting pretty deep in nested with statements. Could we use `contextlib.ExitStatc` instead?